Exception when I try to add a player as a friend after chat refactoring

```
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at javafx.fxml/javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1787)
	at javafx.fxml/javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1670)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.controls/javafx.scene.control.MenuItem.fire(MenuItem.java:465)
	at javafx.controls/com.sun.javafx.scene.control.ContextMenuContent$MenuItemContainer.doSelect(ContextMenuContent.java:1380)
	at javafx.controls/com.sun.javafx.scene.control.ContextMenuContent$MenuItemContainer.lambda$createChildren$12(ContextMenuContent.java:1333)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:218)
	at javafx.base/com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
	at javafx.base/com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
	at javafx.base/com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
	at javafx.base/com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
	at javafx.base/com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
	at javafx.base/javafx.event.Event.fireEvent(Event.java:198)
	at javafx.graphics/javafx.scene.Scene$MouseHandler.process(Scene.java:3876)
	at javafx.graphics/javafx.scene.Scene$MouseHandler.access$1300(Scene.java:3604)
	at javafx.graphics/javafx.scene.Scene.processMouseEvent(Scene.java:1874)
	at javafx.graphics/javafx.scene.Scene$ScenePeerListener.mouseEvent(Scene.java:2613)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:397)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler$MouseEventNotification.run(GlassViewEventHandler.java:295)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleMouseEvent$2(GlassViewEventHandler.java:434)
	at javafx.graphics/com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:389)
	at javafx.graphics/com.sun.javafx.tk.quantum.GlassViewEventHandler.handleMouseEvent(GlassViewEventHandler.java:433)
	at javafx.graphics/com.sun.glass.ui.View.handleMouseEvent(View.java:556)
	at javafx.graphics/com.sun.glass.ui.View.notifyMouse(View.java:942)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:175)
	at java.base/java.lang.Thread.run(Thread.java:844)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:76)
	at jdk.internal.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at javafx.base/com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:275)
	at javafx.fxml/com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:83)
	at javafx.fxml/javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1784)
	... 43 more
Caused by: java.lang.NullPointerException
	at com.faforever.client.chat.ChatUserContextMenuController.onAddFriendSelected(ChatUserContextMenuController.java:261)
	... 54 more
```

--- a/.idea/modules.xml
+++ b/.idea/modules.xml
@@ -2,6 +2,7 @@
 <project version="4">
   <component name="ProjectModuleManager">
     <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/modules/downlords-faf-client.iml" filepath="$PROJECT_DIR$/.idea/modules/downlords-faf-client.iml" />
       <module fileurl="file://$PROJECT_DIR$/downlords-faf-client.iml" filepath="$PROJECT_DIR$/downlords-faf-client.iml" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/webview-patch/downlords-faf-client-webview-patch.iml" filepath="$PROJECT_DIR$/.idea/modules/webview-patch/downlords-faf-client-webview-patch.iml" group="webview-patch" />
       <module fileurl="file://$PROJECT_DIR$/.idea/modules/webview-patch/downlords-faf-client-webview-patch_main.iml" filepath="$PROJECT_DIR$/.idea/modules/webview-patch/downlords-faf-client-webview-patch_main.iml" group="webview-patch" />

--- a/src/main/java/com/faforever/client/chat/ChannelTabController.java
+++ b/src/main/java/com/faforever/client/chat/ChannelTabController.java
@@ -375,7 +375,7 @@ public class ChannelTabController extends AbstractChatTabController {
 
       ChatChannelUser user = listItem.getUser();
 
-      return user != null && user.getUsername().toLowerCase(US).contains(searchString.toLowerCase(US));
+      return listItem.getCategory() != null || user.getUsername().toLowerCase(US).contains(searchString.toLowerCase(US));
     });
   }
 
@@ -429,7 +429,9 @@ public class ChannelTabController extends AbstractChatTabController {
     Arrays.stream(ChatUserCategory.values())
         .filter(chatUserCategory -> !chatUserCategories.contains(chatUserCategory))
         .map(categoriesToUserListItems::get)
-        .forEach(chatUserListItems::removeAll);
+        .flatMap(Collection::stream)
+        .filter(categoryOrChatUserListItem -> categoryOrChatUserListItem.getUser() != null && categoryOrChatUserListItem.getUser().equals(chatUser))
+        .forEach(chatUserListItems::remove);
 
     CategoryOrChatUserListItem listItem = new CategoryOrChatUserListItem(null, chatUser);
     chatUserCategories.stream()
@@ -439,6 +441,7 @@ public class ChannelTabController extends AbstractChatTabController {
 
   private void addToTreeItemSorted(CategoryOrChatUserListItem parent, CategoryOrChatUserListItem child) {
     Platform.runLater(() -> {
+      categoriesToUserListItems.get(parent.getCategory()).add(child);
       for (int index = chatUserListItems.indexOf(parent) + 1; index < chatUserListItems.size(); index++) {
         CategoryOrChatUserListItem otherItem = chatUserListItems.get(index);
 
@@ -483,6 +486,7 @@ public class ChannelTabController extends AbstractChatTabController {
       } else {
         updateUserMessageDisplay(chatUser, "");
       }
+      updateChatUserListItemsForCategories(chatUser);
     };
     socialStatusMessagesListeners.computeIfAbsent(player.getUsername(), i -> new ArrayList<>()).add(listener);
     return new WeakChangeListener<>(listener);

--- a/src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java
+++ b/src/main/java/com/faforever/client/chat/ChatUserContextMenuController.java
@@ -48,7 +48,6 @@ import org.springframework.stereotype.Component;
 import java.lang.invoke.MethodHandles;
 import java.net.URL;
 import java.util.Objects;
-import java.util.Optional;
 
 import static com.faforever.client.chat.ChatColorMode.CUSTOM;
 import static com.faforever.client.fx.WindowController.WindowButtonType.CLOSE;
@@ -95,7 +94,6 @@ public class ChatUserContextMenuController implements Controller<ContextMenu> {
 
   @SuppressWarnings("FieldCanBeLocal")
   private ChangeListener<Player> playerChangeListener;
-  private Optional<Runnable> onSocialStatusChangedListener;
 
   public ChatUserContextMenuController(PreferencesService preferencesService,
                                        PlayerService playerService, ReplayService replayService,
@@ -110,7 +108,6 @@ public class ChatUserContextMenuController implements Controller<ContextMenu> {
     this.joinGameHelper = joinGameHelper;
     this.avatarService = avatarService;
     this.uiService = uiService;
-    onSocialStatusChangedListener = Optional.empty();
   }
 
   public void initialize() {
@@ -258,13 +255,11 @@ public class ChatUserContextMenuController implements Controller<ContextMenu> {
       playerService.removeFoe(player);
     }
     playerService.addFriend(player);
-    onSocialStatusChangedListener.ifPresent(Runnable::run);
   }
 
   public void onRemoveFriendSelected() {
     Player player = getPlayer();
     playerService.removeFriend(player);
-    onSocialStatusChangedListener.ifPresent(Runnable::run);
   }
 
   public void onAddFoeSelected() {
@@ -273,13 +268,11 @@ public class ChatUserContextMenuController implements Controller<ContextMenu> {
       playerService.removeFriend(player);
     }
     playerService.addFoe(player);
-    onSocialStatusChangedListener.ifPresent(Runnable::run);
   }
 
   public void onRemoveFoeSelected() {
     Player player = getPlayer();
     playerService.removeFoe(player);
-    onSocialStatusChangedListener.ifPresent(Runnable::run);
   }
 
   public void onWatchGameSelected() {
@@ -327,9 +320,5 @@ public class ChatUserContextMenuController implements Controller<ContextMenu> {
   @Override
   public ContextMenu getRoot() {
     return chatUserContextMenuRoot;
-  }
-
-  void setOnSocialStatusChangedListener(Optional<Runnable> onSocialStatusChangedListener) {
-    this.onSocialStatusChangedListener = onSocialStatusChangedListener;
   }
 }

--- a/src/main/java/com/faforever/client/chat/ChatUserItemController.java
+++ b/src/main/java/com/faforever/client/chat/ChatUserItemController.java
@@ -108,7 +108,6 @@ public class ChatUserItemController implements Controller<Node> {
   private Tooltip clanTooltip;
   private Tooltip avatarTooltip;
   private Tooltip userTooltip;
-  private Optional<Runnable> onSocialStatusUpdatedListener;
 
   // TODO reduce dependencies, rely on eventBus instead
   public ChatUserItemController(PreferencesService preferencesService, AvatarService avatarService,
@@ -216,7 +215,6 @@ public class ChatUserItemController implements Controller<Node> {
   public void onContextMenuRequested(ContextMenuEvent event) {
     ChatUserContextMenuController contextMenuController = uiService.loadFxml("theme/chat/chat_user_context_menu.fxml");
     contextMenuController.setChatUser(chatUser);
-    contextMenuController.setOnSocialStatusChangedListener(onSocialStatusUpdatedListener);
     contextMenuController.getContextMenu().show(chatUserItemRoot, event.getScreenX(), event.getScreenY());
   }
 
@@ -502,9 +500,5 @@ public class ChatUserItemController implements Controller<Node> {
   public void onMouseExitedAvatarImageView() {
     Tooltip.uninstall(avatarImageView, avatarTooltip);
     avatarTooltip = null;
-  }
-
-  public void setOnSocialStatusUpdatedListener(Optional<Runnable> onSocialStatusUpdatedListener) {
-    this.onSocialStatusUpdatedListener = onSocialStatusUpdatedListener;
   }
 }

--- a/src/main/java/com/faforever/client/chat/ChatUserListCell.java
+++ b/src/main/java/com/faforever/client/chat/ChatUserListCell.java
@@ -4,13 +4,11 @@ import com.faforever.client.theme.UiService;
 import javafx.scene.control.ListCell;
 
 import java.util.Objects;
-import java.util.Optional;
 
 public class ChatUserListCell extends ListCell<CategoryOrChatUserListItem> {
 
   private final ChatUserItemController chatUserItemController;
   private final ChatUserItemCategoryController chatUserCategoryController;
-  private Optional<Runnable> onSocialStatusUpdatedListener;
   private Object oldItem;
 
   public ChatUserListCell(UiService uiService) {
@@ -18,11 +16,6 @@ public class ChatUserListCell extends ListCell<CategoryOrChatUserListItem> {
     chatUserCategoryController = uiService.loadFxml("theme/chat/chat_user_category.fxml");
 
     setText(null);
-  }
-
-  // TODO check whether or not this is meant to be used. Fix accordingly.
-  public void setOnSocialStatusUpdatedListener(Runnable onSocialStatusUpdatedListener) {
-    this.onSocialStatusUpdatedListener = Optional.ofNullable(onSocialStatusUpdatedListener);
   }
 
   @Override
@@ -44,7 +37,6 @@ public class ChatUserListCell extends ListCell<CategoryOrChatUserListItem> {
     if (item.getUser() != null) {
       chatUserCategoryController.setChatUserCategory(null);
       chatUserItemController.setChatUser(item.getUser());
-      chatUserItemController.setOnSocialStatusUpdatedListener(onSocialStatusUpdatedListener);
       setGraphic(chatUserItemController.getRoot());
     } else {
       chatUserItemController.setChatUser(null);

--- a/src/main/java/com/faforever/client/query/LogicalNodeController.java
+++ b/src/main/java/com/faforever/client/query/LogicalNodeController.java
@@ -43,6 +43,9 @@ public class LogicalNodeController implements Controller<Node> {
     logicalOperatorField.setConverter(new StringConverter<>() {
       @Override
       public String toString(LogicalOperator object) {
+        if (object == null) {
+          return "";
+        }
         return i18n.get(object.i18nKey);
       }
 

--- a/src/main/java/com/faforever/client/query/SpecificationController.java
+++ b/src/main/java/com/faforever/client/query/SpecificationController.java
@@ -114,9 +114,12 @@ public class SpecificationController implements Controller<Node> {
     valueField.editableProperty().bind(valueField.visibleProperty());
 
     operationField.setItems(comparisonOperators);
-    operationField.setConverter(new StringConverter<ComparisonOperator>() {
+    operationField.setConverter(new StringConverter<>() {
       @Override
       public String toString(ComparisonOperator object) {
+        if (object == null) {
+          return "";
+        }
         return i18n.get(operatorToI18nKey.get(object));
       }
 

--- a/src/test/java/com/faforever/client/chat/PircBotXChatServiceTest.java
+++ b/src/test/java/com/faforever/client/chat/PircBotXChatServiceTest.java
@@ -840,7 +840,6 @@ public class PircBotXChatServiceTest extends AbstractPlainJavaFxTest {
     when(user.getNick()).thenReturn(chatUser1.getUsername());
 
     connect();
-    firePircBotXEvent(createPrivateMessageEvent(user, message));
 
     CountDownLatch latch = new CountDownLatch(1);
     doAnswer(invocation -> {
@@ -848,7 +847,8 @@ public class PircBotXChatServiceTest extends AbstractPlainJavaFxTest {
       return null;
     }).when(eventBus).post(any(ChatMessageEvent.class));
 
-    latch.await(500, TimeUnit.SECONDS);
+    firePircBotXEvent(createPrivateMessageEvent(user, message));
+    latch.await(100, TimeUnit.SECONDS);
     verify(eventBus).post(any(ChatMessageEvent.class));
   }
 }

